File: languages/swift/cross-language-features.swift
import Foundation
typealias xl = WillyHorizont.Runtime.Xl

@main
struct App {
    static func main() {
        /*
        1. support lambda as value, or has workaround
        */
        let sayHello = { (va: [Any?]) -> Any? in
            var itr = va.makeIterator()
            let callbackFunction = itr.next()!! as! ([Any?]) -> Any?
            print("hello")
            _ = callbackFunction([])
            return nil as Any?
        }
        _ = sayHello([{ (va: [Any?]) -> Any? in
            print("world")
            return nil as Any?
        }])
        let createMultiplier = { (va: [Any?]) -> Any? in
            var itr = va.makeIterator()
            let aa = itr.next()!! as! Int
            return { (va: [Any?]) -> Any? in
                var itr = va.makeIterator()
                let bb = itr.next()!! as! Int
                return aa * bb as Any?
            }
        }
        let multiplyByTwo = createMultiplier([2])
        print("multiply_by_two(10): \((multiplyByTwo! as! ([Any?]) -> Any?)([10]) as! Int)")
        let multiplyByEight = createMultiplier([8])
        print("multiply_by_eight(4): \((multiplyByEight! as! ([Any?]) -> Any?)([4]) as! Int)")
        print("multiply_by_two(8): \((multiplyByTwo! as! ([Any?]) -> Any?)([8]) as! Int)")

        /*
        2. support dynamic-typed value, or has workaround
        */
        let xlList = [
            nil,
            true,
            false,
            "foo",
            0,
            -123,
            123.789,
            -123.789,
            [1, 2, 3],
            ["foo": "bar"],
            { (va: [Any?]) -> Any? in
                var itr = va.makeIterator()
                let aa = itr.next()!! as! Int
                let bb = itr.next()!! as! Int
                return aa * bb as Any?
            }
        ] as [Any?]
        print("xl_list: \(xl.jsonStringify(xlList))")
        print("xl_list: \(xl.jsonStringify(xlList, pretty: true))")
        let xlDict = [
            "xl_none": nil,
            "xl_bool_true": true,
            "xl_bool_false": false,
            "xl_string": "foo",
            "xl_int_positive": 0,
            "xl_int_negative": -123,
            "xl_float_positive": 123.789,
            "xl_float_negative": -123.789,
            "xl_list": [1, 2, 3],
            "xl_dict": ["foo": "bar"],
            "xl_lambda": { (va: [Any?]) -> Any? in
                var itr = va.makeIterator()
                let aa = itr.next()!! as! Int
                let bb = itr.next()!! as! Int
                return aa * bb as Any?
            },
        ] as [String: Any?]
        print("xl_dict: \(xl.jsonStringify(xlDict))")
        print("xl_dict: \(xl.jsonStringify(xlDict, pretty: true))")
    }
}

File: languages/julia/cross-language-features.jl
include("../../runtimes/julia/willyhorizont/runtime/xl.jl")

#=
1. support lambda as value, or has workaround
=#
say_hello = (callback_function) -> begin
    println("hello")
    callback_function()
end
say_hello(() -> begin
    println("world")
end)
create_multiplier = aa -> bb -> aa * bb
multiply_by_two = create_multiplier(2)
println("multiply_by_two(10): $(multiply_by_two(10))")
multiply_by_eight = create_multiplier(8)
println("multiply_by_eight(4): $(multiply_by_eight(4))")
println("multiply_by_two(8): $(multiply_by_two(8))")

#=
2. support dynamic-typed value, or has workaround
=#
xl_list = [
    nothing,
    true,
    false,
    "foo",
    0,
    -123,
    123.789,
    -123.789,
    [1, 2, 3],
    Dict("foo" => "bar"),
    (aa, bb) -> aa * bb,
]
println("xl_list: $(Xl.json_stringify(xl_list))")
println("xl_list: $(Xl.json_stringify(xl_list, pretty=true))")
xl_dict = Dict(
    "xl_none" => nothing,
    "xl_bool_true" => true,
    "xl_bool_false" => false,
    "xl_string" => "foo",
    "xl_int_positive" => 0,
    "xl_int_negative" => -123,
    "xl_float_positive" => 123.789,
    "xl_float_negative" => -123.789,
    "xl_list" => [1, 2, 3],
    "xl_dict" => Dict("foo" => "bar"),
    "xl_lambda" => (aa, bb) -> aa * bb,
)
println("xl_dict: $(Xl.json_stringify(xl_dict))")
println("xl_dict: $(Xl.json_stringify(xl_dict, pretty=true))")
